tests/checkers/unittest_typecheck.py

```diff
@@ -286,6 +287,62 @@ def test_typing_namedtuple_unsubscriptable_object_issue1295(self):
         with self.assertNoMessages():
             self.checker.visit_subscript(subscript)
 
+    def test_typing_option_object_is_subscriptable_issue3882(self):
+        module = astroid.parse(
+            """
+        import typing
+        test = typing.Optional[int]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def test_decorated_by_a_subscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Deco:
+            def __init__(self, f):
+                self.f = f
+
+            def __getitem__(self, item):
+                return item
+        @Deco
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def test_decorated_by_an_unsubscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Deco:
+            def __init__(self, f):
+                self.f = f
+
+        @Deco
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertAddsMessages(
+            Message(
+                "unsubscriptable-object",
+                node=subscript.value,
+                args="decorated",
+                confidence=UNDEFINED,
+            )
+        ):
+            self.checker.visit_subscript(subscript)
+
```

Comment:
Maybe we could add a test with a function decorated by more than one decorator and with one of them that enable the protocol?
What do you think about it?

Reply:
I think we should also test the case where a subscriptable class is decorated by an unsubscriptable decorator.
Something along the lines:
```
    def test_subscriptable_decorated_by_an_unsubscriptable_class_issue3882(self):
        module = astroid.parse(
            """
        class Deco:
            def __init__(self, f):
                self.f = f

        @Deco
        class Decorated:
            def __getitem__(self, item):
                return item

        test = Decorated()[None]
        """
        )
        subscript = module.body[-1].value
        with self.assertNoMessages():
            self.checker.visit_subscript(subscript)
```

Reply:
> Maybe we could add a test with a function decorated by more than one decorator and with one of them that enable the protocol?
>  What do you think about it?

I'm not 100% sure I understand correctly when you say "one of them that enable the protocol", for me it depends too much on the next decorators that can easily "re-drop" the protocol by simply returning a function.

Reply:
I think your example should typically message an error, in this specific case it should warn than `Decorated()`, as being an instance of `Deco` is not callable.

If we then add a `__call__` to Deco that returns something that does *not* implement the protocol, the `Deco` removed the protocol from `Decorated`, and we should say so (but it's probably a whole other level of complexity).

Reply:
Ok, forget the case with multiple decorators.

Reply:
Of forget also my suggestion. It was not pertinent.